lib/stdlib/src/filelib.erl

```diff
@@ -354,15 +354,25 @@ prepare_base(Base0) ->
     lists:reverse(Base2).
 
 do_double_star(Base, [H|T], Patterns, Result0, Mod, Root) ->
-    Full = case Root of
-               false -> filename:join(Base, H);
-               true -> H
+    {Full, Loop} = case Root of
+               false ->
+                    F = filename:join(Base, H),
+                    L = case file:read_link(F) of
+                            {ok, LinkTarget} ->
+                                lists:prefix(filename:split(LinkTarget), filename:split(Base));
```

Comment:
`file:readlink/1` does not canonicalize the returned path. If the target for example is  `../foo` the prefix check will not trigger. I think the traditional UNIX way of detecting symlink loops (or for that matter hard links) is to do a `stat` and then check if you have seen the device and inode before. That will only require _O(n)_ memory, where _n_ is the number of existing files.

On UNIX-like platforms you could probably use `file:read_file_info/1` but I don't know if you will get meaningful inode+device values on Windows.

Reply:
Yes, you are absolutely right about the relative path symlinks. It reminds me another reason why the first proposal here was to not follow symlinks at all.
I don't think hard links can cause any problem.
I'll definitely need more time to come with right solution.